mypy/plugins/re.py

```diff
@@ -0,0 +1,358 @@
+        return None
+    return groups
+
+
+@lru_cache(None)
+def _parse_groups(value: str) -> Groups | re.error:
+    """The 'most important part' of this feature, we parse the regex pattern to discern which groups are optional."""
+    try:
+        p = re.compile(value)
+    except re.error as e:
+        return e
+
+    groups: list[int] = []
+    working = []
+    """
+    1 = capturing group
+    0 = optional capturing group
+    -1 = non-capturing group
+    -2 = optional non-capturing group
+    """
+    depth = 0
+    escape = False
+    union = -1
+    character_set = 0
+    """
+    0 = not in a character set
+    1 = `]` is literal in a character set
+    2 = normal character set
+    """
+    comment = False
+    for i in range(len(value)):
+        if escape:
+            escape = False
+            continue
+        char = value[i]
+        if char == ")" and comment:
+            comment = False
+            continue
+        if char == "\\":
+            if character_set:
+                character_set = 2
+            escape = True
+            continue
+        if comment:
+            continue
+        if char == "^" and character_set == 1:
+            continue
+        if char == "]" and character_set == 2:
+            character_set = 0
+            continue
+        if character_set:
+            character_set = 2
+            continue
+        if char == "[":
+            character_set = 1
+            continue
+        if char == "|" and (union > depth or union == -1):
+            union = depth
+            continue
+        if char not in ("(", ")"):
+            continue
+        if char == "(":
+            depth += 1
+            if value[i + 1] == "?":
+                if value[i + 2 : i + 4] == "P<":
+                    working.append(1)
+                elif value[i + 2] == "!":
+                    working.append(-2)
+                elif value[i + 2] in "#":
+                    comment = True
+                    continue
+                else:
+                    working.append(-1)
+            else:
+                working.append(1)
+            continue
+        if char == ")":
+            depth -= 1
+            if i + 1 < len(value) and value[i + 1] in "*?":
+                for d in range(depth, len(working)):
+                    if working[d] == 1:
+                        working[d] = 0
+            if depth == union - 1 or (depth < len(working) and working[depth] < -1):
+                for d in range(depth + 1, len(working)):
+                    if working[d] == 1:
+                        working[d] = 0
+                union = -1
+            if depth == 0:
+                groups.extend(group for group in working if group > -1)
+                working.clear()
+    if union == 0:
+        for d in range(len(groups)):
+            if groups[d] == 1:
+                groups[d] = 0
+    assert len(groups) == p.groups, "parsed groups differ from compiled groups"
```

Comment:
idk which is which but is it worth including the values for each one in the error message
```suggestion
    assert len(groups) == p.groups, f"parsed groups  ({groups}) differ from compiled groups ({p.groups})"
```

Reply:
what values?

Reply:
the group names

Reply:
yeah, it would have that already